# Homework 4
Quentin Baker

## Problem 1
Deep networks are already clearly on a different trajectory compared to perceptrons and expert systems. This is due to several important factors: modeling, data science, and connectivity. Our models are better -- perceptrons have given way to multi-level neural nets with non-linear activations, and can model successfully problems that stumped perceptrons, most famously the XOR. Data science has improved; we store more data and can access it faster, meaning its possible to train with larger sets. Third, and in my opinion most important, computers are now embedded in all sorts of things and producing meaningful data. We have access to sources of data that would have been unthinkable even a decade ago, and this is just the opening salvo. Regardless of opinion on this data collection, it is undisputable that it means a lot for NNs.

## Problem 2
![Page 1](Part2Page1.jpg)

![Page 2](Part2Page2.jpg)

## Problem 3

Create a sequential NN for our classifier (from kvlinden-courses/cs344-code/unit09)

In [1]:
from keras import layers
from keras import models
model = models.Sequential()

# Configure a convnet with 3 layers of convolutions and max pooling.
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Add layers to flatten the 2D image and then do a 10-way classification.
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

model.summary()

Using TensorFlow backend.


Instructions for updating:
Colocations handled automatically by placer.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dens


Import the dataset and build a basic Keras structure (from https://keras.io/datasets/#fashion-mnist-database-of-fashion-articles)

In [8]:
from keras.datasets import fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

According to the [MNIST repo][mnist-git], the data is split between 60k training and 10k test examples. We can confirm this ourselves:

[mnist-git]:https://github.com/zalandoresearch/fashion-mnist

In [3]:
print('training set:\t%s images' % len(train_images))
print('test set:\t%s images' % len(test_images))

training set:	60000 images
test set:	10000 images


Now before continuing, I'm going to reshape the data. As this is similar to the classic MNIST digits set (28x28 pixel grayscale images) I will reshape in a similar manner, then convert the set to a categorical model:

In [10]:
from keras.utils import to_categorical

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

Now for the secret sauce: Hyperparams and training! I chose binary crossentropy for the loss model as I found in testing it seemed to perform better than categorical.

In [11]:
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=50)
model.evaluate(test_images, test_labels)

Epoch 1/5
60000/60000 [==============================] - 36s 596us/step - loss: 0.0389 - acc: 0.9846
Epoch 2/5
60000/60000 [==============================] - 42s 705us/step - loss: 0.0361 - acc: 0.9858
Epoch 3/5
60000/60000 [==============================] - 42s 706us/step - loss: 0.0339 - acc: 0.9869
Epoch 4/5
60000/60000 [==============================] - 42s 708us/step - loss: 0.0317 - acc: 0.9878
Epoch 5/5
10000/10000 [==============================] - 2s 220us/step


[0.052288673339784146, 0.9822200072288513]

I'm not sure if I had a fluke, but I ran this several times and saw >= 98% accuracy each time! I actually *lowered* the batch size since it is performing so well.